File: src/main/java/五毒/第三周/第二遍/combination.java
package 五毒.第三周.第二遍;

import java.util.ArrayList;
import java.util.List;

public class combination {
    public static List<List<Integer>> combine(int n, int k) {
        List<List<Integer>> result = new ArrayList<>();
        if(n<k){
            return result;
        }
        int depth = 1;
        List<Integer> list = new ArrayList<>();
        dfs(result,depth,list,n,k);
        return result;
    }

    private static void dfs(List<List<Integer>> result, int depth, List<Integer> list,int n, int k) {
        if(list.size()==k){
            result.add(new ArrayList<>(list));
            return;
        }

        for(int i = depth;i<=n;i++){
            list.add(i);
            dfs(result,i+1,list,n,k);
            list.remove(list.size()-1);
        }
    }

    public static void main(String[] args) {
        System.out.println( combine(4,2));
    }

}

File: src/main/java/五毒/第三周/第二遍/combination.go
package main

import "fmt"

func combine(n int, k int) [][]int {
	result := make([][]int, 0)
	list := make([]int, 0)
	depth := 1
	dfs(&result, &list, depth, n, k) //传地址才能深拷贝
	return result
}

func dfs(result *[][]int, list *[]int, depth int, n int, k int) {
	if len(*list) == k {
		newList := make([]int, 2)
		copy(newList, *list)
		fmt.Println(newList)
		*result = append(*result, newList)
		return
	}
	var i int
	for i = depth; i <= n; i++ {
		*list = append(*list, i)
		dfs(result, list, i+1, n, k)
		*list = (*list)[0 : len(*list)-1]
	}
}

func main() {
	fmt.Println(combine(4, 2))
}
